lib/async_hooks.js

```diff
@@ -342,24 +345,28 @@ function emitInitS(asyncId, type, triggerAsyncId, resource) {
   }
 }
 
-
-function emitBeforeN(asyncId) {
-  processing_hook = true;
-  // Use a single try/catch for all hook to avoid setting up one per iteration.
-  try {
-    for (var i = 0; i < active_hooks_array.length; i++) {
-      if (typeof active_hooks_array[i][before_symbol] === 'function') {
-        active_hooks_array[i][before_symbol](asyncId);
+function hookFactory(symbol) {
+  // Called from native. The asyncId stack handling is taken care of there
+  // before this is called.
+  return function(asyncId) {
```

Comment:
Have we verified how this'll change the stack trace from any of these? If V8 can properly infer the name based on the variable name then great.